Prefer maybe_multiply when multiplying by a constant

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* #68522
* #69909
* #69908
* __->__ #68185
* #68184
* #68183
* #67014
* #66934
* #67789
* #66933

As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

lezcano committed:
Prefer maybe_multiply when multiplying by a constant

As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

Lezcano commented:
Could you have a look at this @albanD ?

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

mruberry commented on tools/autograd/derivatives.yaml:
```diff
@@ -1348,19 +1348,19 @@
 - name: rsub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
-  self: handle_r_to_c(other.scalar_type(), -grad * alpha.conj())
-  other: handle_r_to_c(self.scalar_type(), grad)
+  self: handle_r_to_c(self.scalar_type(), maybe_multiply(-grad, alpha.conj()))
```
Would you elaborate on this change? Is there a test this can add that would have failed without it and now passes?

Lezcano replied:
I reckon that if you do rsub between a float and a complex tensor, this test should fail. otoh, we did not generate such samples in the OpInfo. Now, this kind of samples are generated in the next PR of this stack.

mruberry reviewed (approved):
Awesome

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

albanD reviewed (approved):
Thanks

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

mruberry commented:
Hmm... are these errors related?

```
======================================================================
ERROR [0.024s]: test_lu_solve_batched_many_batches_cuda_complex128 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1050, in wrapper
    fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7855, in test_lu_solve_batched_many_batches
    run_test((5, 65536), (65536, 5, 10))
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7851, in run_test
    x = torch.lu_solve(b, LU_data, LU_pivots)
RuntimeError: Number of pivots per batch should be same as the dimension of the matrix

======================================================================
ERROR [0.009s]: test_lu_solve_batched_many_batches_cuda_complex64 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1050, in wrapper
    fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7855, in test_lu_solve_batched_many_batches
    run_test((5, 65536), (65536, 5, 10))
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7851, in run_test
    x = torch.lu_solve(b, LU_data, LU_pivots)
RuntimeError: Number of pivots per batch should be same as the dimension of the matrix

======================================================================
ERROR [0.008s]: test_lu_solve_batched_many_batches_cuda_float32 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1050, in wrapper
    fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7855, in test_lu_solve_batched_many_batches
    run_test((5, 65536), (65536, 5, 10))
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 7851, in run_test
    x = torch.lu_solve(b, LU_data, LU_pivots)
RuntimeError: Number of pivots per batch should be same as the dimension of the matrix

======================================================================
ERROR [0.013s]: test_lu_solve_batched_many_batches_cuda_float64 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
======================================================================
ERROR [0.076s]: test_old_solve_batched_many_batches_cuda_float32 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1050, in wrapper
    fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 3732, in test_old_solve_batched_many_batches
    x, _ = torch.solve(b, A)
RuntimeError: Incompatible matrix sizes for solve: each A matrix is 256 by 256 but each b matrix is 5 by 1

======================================================================
ERROR [0.495s]: test_old_solve_batched_many_batches_cuda_float64 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1050, in wrapper
    fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_linalg.py", line 3732, in test_old_solve_batched_many_batches
    x, _ = torch.solve(b, A)
RuntimeError: Incompatible matrix sizes for solve: each A matrix is 256 by 256 but each b matrix is 5 by 1
```

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).

lezcano committed:
Update on "Prefer maybe_multiply when multiplying by a constant"


As per title

We also fix the first input to `handle_r_to_c` fo `rsub`as it was
flipped for the two inputs.

Differential Revision: [D32684855](https://our.internmc.facebook.com/intern/diff/D32684855)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684855).
